File: Letter_Combinations_of_a_Phone_Number.java
/**
 * Problem Link: https://leetcode.com/problems/letter-combinations-of-a-phone-number/
 * 
 * Iterate all of the possible solutions.
 */

public class Solution {
    private HashMap<Integer, String> map;
    private List<String> list;
    public List<String> letterCombinations(String digits) {
        map = new HashMap<Integer, String>();
        map.put(0, "");
        map.put(1, "");
        map.put(2, "abc");
        map.put(3, "def");
        map.put(4, "ghi");
        map.put(5, "jkl");
        map.put(6, "mno");
        map.put(7, "pqrs");
        map.put(8, "tuv");
        map.put(9, "wxyz");
        list = new ArrayList<String>();
        combine(digits, 0, "");
        return list;
    }
    
    private void combine(String digit, int current, String ret) {
        if(current == digit.length()) {
            if(ret.length() > 0) list.add(ret);
            return;
        }
        String s = map.get(digit.charAt(current) - '0');
        for(int i=0; i<s.length(); i++) {
            char c = s.charAt(i);
            combine(digit, current + 1, ret + c);
        }
    }
}

File: Swift/letter-combinations-of-a-phone-number.swift
/**
 * Problem Link: https://leetcode.com/problems/letter-combinations-of-a-phone-number/
 *
 *
 */

class Solution {
    let dict = [
        1:["1"],
        2:["a", "b", "c"],
        3:["d", "e", "f"],
        4:["g", "h", "i"],
        5:["j", "k", "l"],
        6:["m", "n", "o"],
        7:["p", "q", "r", "s"],
        8:["t", "u", "v"],
        9:["w", "x", "y", "z"],
        0:["0"]
    ]
    func letterCombinations(_ digits: String) -> [String] {
        return Array(comb(digits.characters.map({Int(String($0))!})))
    }
    
    private func comb(_ digits: [Int]) -> Set<String> {
        var rSet = Set<String>()
        if digits.count == 0 {
            return rSet
        }
        if digits.count == 1 {
            rSet = Set<String>(dict[digits[0]]!)
            return rSet
        }
        let subSet = comb(Array(digits.dropFirst(1)))
        for c in dict[digits[0]]! {
            for s in subSet {
                rSet.insert(c + s)
            }
        }
        return rSet
    }
}
